File: 700/714 - Best Time to Buy and Sell Stock with Transaction Fee.java
class Solution {
    public int maxProfit(int[] prices, int fee) {
        int len = prices.length;
        int[] hold = new int[len], sell = new int[len];
        
        hold[0] = -prices[0];
        
        for (int i = 1; i < len; ++i) {
            sell[i] = Math.max(sell[i - 1], hold[i - 1] + prices[i] - fee);
            hold[i] = Math.max(hold[i - 1], sell[i - 1] - prices[i]);
        }
        
        return sell[len - 1];
    }
}

File: csharp/714.best-time-to-buy-and-sell-stock-with-transaction-fee.cs
/*
 * @lc app=leetcode id=714 lang=csharp
 *
 * [714] Best Time to Buy and Sell Stock with Transaction Fee
 */

// @lc code=start
public class Solution {
    public int MaxProfit(int[] prices, int fee) {
        int h = -prices[0], s = 0, n = prices.Length;
        foreach (int price in prices) {
            int os = s;
            s = Math.Max(s, h + price - fee);
            h = Math.Max(h, os - price);
        }
        return s;
    }
}
// @lc code=end

